mmcv/ops/csrc/common/cuda/rans_cuda_kernel.cuh

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) OpenMMLab. All rights reserved.
+#ifndef RANS_CUDA_KERNEL_CUH
+#define RANS_CUDA_KERNEL_CUH
+
+#ifdef MMCV_USE_PARROTS
+#include "parrots_cuda_helper.hpp"
+#else
+#include "pytorch_cuda_helper.hpp"
+#endif
+
+#include "utils/rans/rans.hpp"
+
+#define BYTES_OF(x) sizeof(decltype(x))
+#define CUDA_MALLOC(x, num) cudaMalloc((void**)&x, num* BYTES_OF(*x))
+#define CUDA_MEMSET(x, num, value) cudaMemset(x, value, num* BYTES_OF(*x))
+#define CUDA_TO_CPU(x_cuda, x_cpu, num) \
+  cudaMemcpy(x_cpu, x_cuda, num* BYTES_OF(*x_cuda), cudaMemcpyDeviceToHost)
+#define CPU_TO_CUDA(x_cpu, x_cuda, num) \
+  cudaMemcpy(x_cuda, x_cpu, num* BYTES_OF(*x_cuda), cudaMemcpyHostToDevice)
+#define CREATE_TENSOR(shape, type, target_device) \
+  zeros(shape, TensorOptions().dtype(type).device(target_device));
+#define CUDA_FREE(x) cudaFree(x)
+
+struct RansCUDAKernelResult {
+  uint32_t raw_value;
+  bool bypass_mode;
+};
+
+// ---------------------------- Encoding -------------------------------------
+__global__ void rans_encode_with_indexes_cuda_kernel(
+    const int* symbols, const int* indexes, const int* cdfs,
+    const int* cdfs_sizes, const int* offsets, const int num_symbols,
+    const int num_cdfs, const int max_cdf_bin_size, RansSymbol* rans_symbols,
+    RansCUDAKernelResult* results) {
+  int i = blockDim.x * blockIdx.x + threadIdx.x;
+  if (i < num_symbols) {
+    const size_t cdf_idx = indexes[i];
+    assert(cdf_idx < num_cdfs);
+
+    const int* cdf = cdfs + cdf_idx * max_cdf_bin_size;
+
+    const int32_t max_value = cdfs_sizes[cdf_idx] - 2;
+    assert(max_value >= 0 && (max_value + 1) < max_cdf_bin_size);
+
+    int32_t value = symbols[i] - offsets[cdf_idx];
+
+    uint32_t raw_val = 0;
+    if (value < 0) {
+      raw_val = -2 * value - 1;
+      value = max_value;
+    } else if (value >= max_value) {
+      raw_val = 2 * (value - max_value);
+      value = max_value;
+    }
+
+    assert(value >= 0 && value < max_value + 1);
+
+    rans_symbols[i].start = cdf[value];
```

Comment:
load and store from global memory is heavy. Please cache the value in variable, do the compute and write it to the global memory.